app/src/main/java/org/wikipedia/language/LanguagesListViewModel.kt

```diff
@@ -0,0 +1,82 @@
+package org.wikipedia.language
+
+import android.content.Context
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import org.apache.commons.lang3.StringUtils
+import org.wikipedia.R
+import org.wikipedia.WikipediaApp
+import org.wikipedia.dataclient.ServiceFactory
+import org.wikipedia.dataclient.mwapi.SiteMatrix
+import org.wikipedia.util.Resource
+import java.util.*
+
+class LanguagesListViewModel : ViewModel() {
+
+    val siteListData = MutableLiveData<Resource<List<SiteMatrix.SiteInfo>>>()
+    private val suggestedLanguageCodes = WikipediaApp.getInstance().language().remainingAvailableLanguageCodes
+    private val nonSuggestedLanguageCodes = WikipediaApp.getInstance().language()
+        .appMruLanguageCodes.toMutableList().also { it.removeAll(suggestedLanguageCodes) }
+
+    init {
+        fetchData()
+    }
+
+    private fun fetchData() {
+        viewModelScope.launch {
+            withContext(Dispatchers.IO) {
+                val siteMatrix = ServiceFactory.get(WikipediaApp.getInstance().wikiSite).getSiteMatrix()
+                val sites = SiteMatrix.getSites(siteMatrix)
+                siteListData.postValue(Resource.Success(sites))
+            }
+        }
+    }
+
+    fun getListBySearchTerm(context: Context, searchTerm: String?): List<LanguageListItem> {
+        val results = mutableListOf<LanguageListItem>()
+        val filter = StringUtils.stripAccents(searchTerm.orEmpty()).lowercase(Locale.getDefault())
+        var first = true
+        for (code in suggestedLanguageCodes) {
+            val localizedName = StringUtils.stripAccents(WikipediaApp.getInstance().language().getAppLanguageLocalizedName(code).orEmpty())
+            val canonicalName = StringUtils.stripAccents(getCanonicalName(code).orEmpty())
+            if (filter.isEmpty() || code.contains(filter) ||
+                localizedName.lowercase(Locale.getDefault()).contains(filter) ||
+                canonicalName.lowercase(Locale.getDefault()).contains(filter)) {
+                if (first) {
+                    results.add(LanguageListItem(context.getString(R.string.languages_list_suggested_text), true))
+                    first = false
+                }
+                results.add(LanguageListItem(code))
+            }
```

Comment:
Is it possible to make this logic a method? Looks like it has pretty much similar to the logic below.

Reply:
correct.